From: Julian Wiedmann <jwi@linux.vnet.ibm.com>
Subject: s390/qeth: use ip_lock for hsuid configuration
Patch-mainline: v4.8-rc8
Git-commit: 016930b88a1d6eb6e6b3287d593e13ca06986acc
References: bnc#1070825, LTC#161871

Description:  qeth: allow hsuid configuration
Symptom:      'hsuid' parameter of HiperSockets device can't be set
Problem:      The 'hsuid' translates to a link-local IPv6 address.
              The hsuid can only be set while the card is DOWN,
              but while in this state the IP registration command fails.
Solution:     Delay the IP registration for hsuid until the card can
              send IP registration commands.
              Also apply proper locking when modifying the card's IP
              address table.
Reproduction: Attempt to set a 'hsuid' while card is in DOWN state.

Upstream-Description:

              s390/qeth: use ip_lock for hsuid configuration

              qeth_l3_dev_hsuid_store() changes the ip hash table, which
              requires the ip_lock.

              Signed-off-by: Ursula Braun <ubraun@linux.vnet.ibm.com>
              Signed-off-by: David S. Miller <davem@davemloft.net>


Signed-off-by: Julian Wiedmann <jwi@linux.vnet.ibm.com>
Acked-by: Hannes Reinecke <hare@suse.com>
---
 drivers/s390/net/qeth_l3_sys.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/drivers/s390/net/qeth_l3_sys.c
+++ b/drivers/s390/net/qeth_l3_sys.c
@@ -297,7 +297,9 @@ static ssize_t qeth_l3_dev_hsuid_store(s
 		addr->u.a6.pfxlen = 0;
 		addr->type = QETH_IP_TYPE_NORMAL;
 
+		spin_lock_bh(&card->ip_lock);
 		qeth_l3_delete_ip(card, addr);
+		spin_unlock_bh(&card->ip_lock);
 		kfree(addr);
 	}
 
@@ -329,7 +331,10 @@ static ssize_t qeth_l3_dev_hsuid_store(s
 		addr->type = QETH_IP_TYPE_NORMAL;
 	} else
 		return -ENOMEM;
+
+	spin_lock_bh(&card->ip_lock);
 	qeth_l3_add_ip(card, addr);
+	spin_unlock_bh(&card->ip_lock);
 	kfree(addr);
 
 	return count;
